### ISSUE TITLE
undetected `possibly unbound variable` in a list comprehension

### ISSUE THREAD

**Issue by dybi** [microsoft/pyright#6685]
When there is a potentially unbound variable, i.e. due to conditional import `pyright` correctly notices when I try to use the conditionally imported symbol - unless the usage is within a list comprehension. Please see the below code (simplified)

```python
import platform

if platform.system() == "Windows":
    import os.path


def get_dirs(dst_roots, file_path):
    if os.path.islink(file_path):
        print(f"{file_path=} is junction")
    return [dst_root for dst_root in dst_roots if os.path.isdir(os.path.dirname(dst_root))]
```

There should be two errors - one about possibly unbound `os` in `if` statement and one about possibly unbound `os` in list comprehension. The actual result is just the first one:
![image](https://github.com/microsoft/pyright/assets/36961162/5786e8ae-cfe6-4d57-bfe3-06bdfb329e89)

**Comment by erictraut** [comment]
Thanks for the bug report. The previous logic for detecting unbound variables was rather conservative. It supported captured variables but only if the capture didn't cross the boundary of a class or comprehension scope. I've extended it to support all scope types. This has broader implications and produced a number of [new errors](https://github.com/microsoft/pyright/pull/6693) in existing typed code bases, and some of these may be seen as false positives. I think the change is justified, however, because it's more consistent.

This will be included in the next release.

**Comment by dybi** [comment]
Glad to help. Thanks for you work! :+1:

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.340, which I just published. It will be included in a future release of pylance.
